core/src/common/empty_block/inherents/providers.rs

```diff
@@ -0,0 +1,167 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//! TODO: Docs
+
+use std::{str::FromStr, time::Duration};
+
+use itertools::Itertools;
+use parity_scale_codec::Encode;
+use sp_consensus_aura::{Slot, SlotDuration, AURA_ENGINE_ID};
+use sp_consensus_babe::{
+    digests::{PreDigest, SecondaryPlainPreDigest},
+    BABE_ENGINE_ID,
+};
+use sp_inherents::InherentData;
+use sp_runtime::{
+    traits::{Block as BlockT, HashingFor},
+    Digest, DigestItem,
+};
+use sp_state_machine::TestExternalities;
+use sp_std::prelude::*;
+use strum::IntoEnumIterator;
+use strum_macros::{Display, EnumIter};
+
+use crate::common::empty_block::inherents::custom_idps;
+
+/// Trait for providing the inherent data and digest items for block construction.
+pub trait InherentProvider<B: BlockT> {
+    type Err;
+
+    fn get_inherent_providers_and_pre_digest(
+        &self,
+        maybe_parent_info: Option<(InherentData, Digest)>,
+        parent_header: B::Header,
+        ext: &mut TestExternalities<HashingFor<B>>,
+    ) -> InherentProviderResult<Self::Err>;
+}
+
+// Clippy asks that we abstract the return type because it's so long
+type InherentProviderResult<Err> =
+    Result<(Box<dyn sp_inherents::InherentDataProvider>, Vec<DigestItem>), Err>;
+
+/// Classes of [`InherentProvider`] avaliable.
+///
+/// Currently only Smart is implemented. New implementations may be added if Smart is not suitable
+/// for some edge cases.
+#[derive(Debug, Clone, clap::Parser, EnumIter, Display, Copy)]
+pub enum ProviderVariant {
+    /// Smart chain varient will automatically adjust provided inherents based on the given
+    /// externalities.
+    Smart,
```

Comment:
Is it smarter than something like `All`?

Reply:
I chose "smart" because the custom IDPs detect what type of chain by inspecting storage and adjust behavior accordingly. "All" would be better if it just blindly applies all of them.